File: week 1 - C/credit.c
#include <cs50.h>
#include <stdio.h>

bool validate_card(long num);
void print_card_type(long number);

int main(void)
{
    long number = get_long("Input Card Number: ");
    if (validate_card(number))
    {
        print_card_type(number);
    }
    else
    {
        printf("INVALID\n");
    }
}

// validate card number
bool validate_card(long num)
{
    int sum = 0;
    int digit_count = 0;

    for (long n = num; n > 0; n /= 10)
    {
        int digit = n % 10;

        if (digit_count % 2 == 0) // If it is an odd number of digits (counting from the right)
        {
            sum += digit;
        }
        else
        {
            int doubled = digit * 2;
            sum += (doubled / 10) + (doubled % 10);
        }

        digit_count++;
    }

    return sum % 10 == 0;
}

// Determine and output the credit card type
void print_card_type(long number)
{
    int length = 0; // Get the length of the credit card number
    long temp = number;

    while (temp > 0)
    {
        temp /= 10;
        length++;
    }

    long first_two_digits = number;
    while (first_two_digits >= 100)
    {
        first_two_digits /= 10;
    }

    if (length == 15 && (first_two_digits == 34 || first_two_digits == 37))
    {
        printf("AMEX\n");
    }
    else if (length == 16 && (first_two_digits >= 51 && first_two_digits <= 55))
    {
        printf("MASTERCARD\n");
    }
    else if ((length == 13 || length == 16) && (first_two_digits / 10 == 4))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: Week6-Python/credit.py
def validate_card(num):
    sum = 0
    digit_count = 0

    # Loop through each digit of the number from right to left
    while num > 0:
        digit = num % 10

        # If it's an odd position (from the right), just add the digit
        if digit_count % 2 == 0:
            sum += digit
        else:
            # If it's an even position, double the digit and add both digits
            doubled = digit * 2
            sum += (doubled // 10) + (doubled % 10)

        num //= 10  # Move to the next digit
        digit_count += 1

    # If the sum is divisible by 10, the card is valid
    return sum % 10 == 0


def print_card_type(number):
    length = len(str(number))  # Get the length of the credit card number
    first_two_digits = int(str(number)[:2])  # Get the first two digits as an integer

    if length == 15 and (first_two_digits == 34 or first_two_digits == 37):
        print("AMEX")
    elif length == 16 and (51 <= first_two_digits <= 55):
        print("MASTERCARD")
    elif (length == 13 or length == 16) and (first_two_digits // 10 == 4):
        print("VISA")
    else:
        print("INVALID")


def main():
    number = int(input("Input Card Number: "))

    if validate_card(number):
        print_card_type(number)
    else:
        print("INVALID")


if __name__ == "__main__":
    main()
